### PR TITLE
Add Forward Pass Logits Verification for the Checkpoint Validation Pipeline

### PR DESCRIPTION
**Pull request by olufiyin19**
# Description

This PR introduces the third phase of the automated post checkpoint conversion validation pipeline. Building upon the static shape and mock tensor tracing established in PR #4462, this PR ensures that converted checkpoints (e.g. Gemma, Llama, Deepseek, Mistral, Qwen) produce numerically accurate logits during an actual forward pass.

The validation is broken into two sequential subtasks:

1. Forward Pass Logit Verification (`forward_pass_validator.py`): Utilizes MaxText's `tests/utils/forward_pass_logit_checker.py` in an automated orchestration layer that executes a forward pass on converted checkpoints, comparing output logits against golden reference distributions and standardizing the result into a JSON schema (`run_name_forward_pass_validation_report.json`) uploaded to Google Cloud Storage.

2. Bidirectional Checkpoint Compatibility & Tokenizer Robustness: Dynamically detects checkpoint formats (linen vs nnx) on-the-fly and transparently maps layer sequence hierarchies, layer norm naming, and tokenizer padding defaults during checkpoint restoration so that converted weights load cleanly without premature buffer deletion or tokenization crashes.

# Key Implementation Details

1. NNX ↔ Linen Key Translation & Dynamic Format Detection
* Implemented `_rename_nnx_linen_keys` to recursively traverse checkpoint trees, dictionaries (`nnx.State`), and lists/tuples to convert between NNX layer naming conventions (e.g., `layers_0`, `layers_1`) and Linen sequence dictionaries (`layers: {'0': ..., '1': ...}`).
* Added dynamic checkpoint format detection inside an Orbax restoration monkeypatch (`_monkeypatched_restore`). Before translating keys, the validator inspects whether the checkpoint stored on GCS uses Linen conventions or is already in NNX format (e.g., unscanned Qwen3-8b checkpoints), preventing premature buffer deletion or key collision errors during state restoration. 

2. SequenceKey Resolution & Shape Mismatch Monkeypatch

* Monkeypatched `model_creation_utils._fix_restore_args_for_shape_mismatch` (`_lookup_stored_meta`) to resolve a bug where MaxText could not map `SequenceKey` list indices (e.g.,` layers.0`) to string keys (`layers_0`) when restoring Linen checkpoints into NNX models.

3. Tokenizer Robustness & Pad-Token Injection

* Monkeypatched `transformers.AutoTokenizer.from_pretrained` (`_monkeypatched_from_pretrained`) to automatically assign a fallback pad_token_id (defaulting to `eos_token` or `0`) when loading tokenizers for architectures that omit an explicit padding token in their `tokenizer_config.json`. This prevents `ValueError` exceptions during batch tokenization in the forward pass.

4. Standardized JSON Reporting & Airflow Integration

* Encapsulated logit checker execution, capturing subprocess stdout/stderr and Python tracebacks into a structured `run_name_forward_pass_validation_report.json`.
* Automatically uploads the report to `report_gcs_dir`, enabling Airflow fail-fast pipelines to inspect logit fidelity and halt downstream decoding or training tasks if numerical divergence is detected.

# Tests

Verified forward-pass logit verification, tokenizer pad-token injection, and bidirectional NNX ↔ Linen key restoration across five distinct architectures. The script correctly parses Airflow overrides, resolves custom GCS tokenizer paths, and validates logit output distributions on TPU and CPU meshes.

<details>
<summary><b>1. Qwen3-14b</b></summary>

**Airflow Trigger JSON:**
```json
{
    "checkpoint_gcs_path": "gs://maxtext-model-checkpoints/qwen3-14b/scanned/2026-04-12/0/items",
    "hf_model_path": "Qwen/Qwen3-14B",
    "hf_token": "***",
    "maxtext_branch": "feature/checkpoint-validation-clean",
    "maxtext_commit_hash": null,
    "maxtext_model_name": "qwen3-14b",
    "maxtext_overrides": {
        "attention": "dot_product",
        "debug_tensors": true,
        "max_target_length": 4096,
        "per_device_batch_size": 4,
        "scan_layers": true,
        "tokenizer_path": "Qwen/Qwen3-14B",
        "tokenizer_type": "huggingface"
    },
    "report_gcs_dir": "gs://maxtext-validation-agent-reports/",
    "run_name": "qwen14b-context-stress-test"
}
```

**Forward Pass Validation Task Execution Log:**
https://paste.googleplex.com/6189524110671872
</details>  

<details>
<summary><b>2. Qwen3-8b</b></summary>

**Airflow Trigger JSON:**
```json
{
    "checkpoint_gcs_path": "gs://maxtext-model-checkpoints/qwen3-8b/unscanned/0/items",
    "hf_config_url": "https://huggingface.co/Qwen/Qwen3-8B/raw/main/config.json",
    "hf_model_path": "Qwen/Qwen3-8B",
    "hf_ref_code_url": "https://raw.githubusercontent.com/huggingface/transformers/main/src/transformers/models/qwen3/modeling_qwen3.py",
    "hf_token": "***",
    "maxtext_branch": "feature/checkpoint-validation-clean",
    "maxtext_commit_hash": null,
    "maxtext_model_name": "qwen3-8b",
    "maxtext_overrides": {
        "attention": "dot_product",
        "debug_tensors": true,
        "max_prefill_predict_length": 16,
        "max_target_length": 128,
        "per_device_batch_size": 4,
        "prompt": "I love to ",
        "scan_layers": false,
        "tokenizer_path": "Qwen/Qwen3-8B",
        "tokenizer_type": "huggingface"
    },
    "report_gcs_dir": "gs://maxtext-validation-agent-reports/",
    "run_name": "master-qwen3-8b-full-run"
}
```

**Forward Pass Validation Task Execution Log:**
https://paste.googleplex.com/5442362154287104
</details>  

<details>
<summary><b>3. Gemma3-4b</b></summary>

**Airflow Trigger JSON:**
```json
{
    "checkpoint_gcs_path": "gs://maxtext-model-checkpoints/gemma3-4b/2025-03-18-19-03/unscanned/checkpoints/0/items",
    "hf_config_url": "https://huggingface.co/google/gemma-3-4b-it/raw/main/config.json",
    "hf_model_path": "google/gemma-3-4b-it",
    "hf_ref_code_url": "https://raw.githubusercontent.com/huggingface/transformers/main/src/transformers/models/gemma3/modeling_gemma3.py",
    "hf_token": "***",
    "maxtext_branch": "feature/checkpoint-validation-clean",
    "maxtext_commit_hash": null,
    "maxtext_model_name": "gemma3-4b",
    "maxtext_overrides": {
        "attention": "dot_product",
        "debug_tensors": true,
        "max_prefill_predict_length": 16,
        "max_target_length": 128,
        "per_device_batch_size": 4,
        "prompt": "I love to ",
        "scan_layers": false,
        "tokenizer_path": "google/gemma-3-4b-it",
        "tokenizer_type": "huggingface"
    },
    "report_gcs_dir": "gs://maxtext-validation-agent-reports/",
    "run_name": "master-gemma3-4b-full-run"
}
```

**Forward Pass Validation Task Execution Log:**
https://paste.googleplex.com/6171948567625728
</details>  

<details>
<summary><b>4. Mistral-7b</b></summary>

**Airflow Trigger JSON:**
```json
{
    "checkpoint_gcs_path": "gs://maxtext-model-checkpoints/mistral-7b/2025-01-23-19-04/unscanned/checkpoints/0/items",
    "hf_config_url": "https://huggingface.co/mistralai/Mistral-7B-v0.1/raw/main/config.json",
    "hf_model_path": "mistralai/Mistral-7B-v0.1",
    "hf_ref_code_url": "https://raw.githubusercontent.com/huggingface/transformers/main/src/transformers/models/mistral/modeling_mistral.py",
    "hf_token": "***",
    "maxtext_branch": "feature/checkpoint-validation-clean",
    "maxtext_commit_hash": null,
    "maxtext_model_name": "mistral-7b",
    "maxtext_overrides": {
        "attention": "dot_product",
        "debug_tensors": true,
        "dtype": "bfloat16",
        "max_prefill_predict_length": 16,
        "max_target_length": 128,
        "per_device_batch_size": 4,
        "prompt": "I love to ",
        "scan_layers": false,
        "tokenizer_path": "mistralai/Mistral-7B-v0.1",
        "tokenizer_type": "huggingface"
    },
    "report_gcs_dir": "gs://maxtext-validation-agent-reports/",
    "run_name": "master-mistral-7b-full-run"
}
```

**Forward Pass Validation Task Execution Log:**
https://paste.googleplex.com/5991767576018944
</details>  

# Stack
2nd PR in stack. Depends on the preceding upstream PR: #4462 

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Comment by olufiyin19** [comment]
@gemini-cli /review

**Comment by olufiyin19** [comment]
@gemini-cli /review

File: src/maxtext/experimental/agent/ckpt_validation_pipeline/__init__.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Checkpoint Validation Agent Package.
+Used to verify and report the status of converted model checkpoints.
+"""
+
+from maxtext.experimental.agent.ckpt_validation_pipeline import layer_metrics
```
[the remark this answers is not in the document]
Comment (olufiyin19) [reply src/maxtext/experimental/agent/ckpt_validation_pipeline/__init__.py:20]: Accounted for later in the PR stack and shouldn't be a problem since we're only merging the last pr in stack

File: src/maxtext/experimental/agent/ckpt_validation_pipeline/forward_compile_validator.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+  parser.add_argument("--maxtext_model_name", type=str, default="", help="MaxText model configuration name")
+
+  args, _overrides = parser.parse_known_args()
+  report_gcs_dir = args.report_gcs_dir
+
+  run_name_override = "unknown"
+  for override in _overrides:
+    if override.startswith("run_name="):
+      run_name_override = override.split("=", 1)[1]
+      break
```
[the remark this answers is not in the document]
Comment (olufiyin19) [reply src/maxtext/experimental/agent/ckpt_validation_pipeline/forward_compile_validator.py:125]: Addressed later on in the stack

File: src/maxtext/experimental/agent/ckpt_validation_pipeline/forward_pass_validator.py [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+
+  # generate report
+  report = {
+      "run_name": run_name,
+      "model": internal_model_name,
+      "status": "SUCCESS" if returncode == 0 else "FAILED",
+      "success": returncode == 0,
+      "stderr": (stderr_str if returncode != 0 else "Success"),
+      "stdout": (stdout_str if returncode != 0 else "Success"),
+      "checkpoint_used": checkpoint_path,
```
[the remark this answers is not in the document]
Comment (olufiyin19) [reply src/maxtext/experimental/agent/ckpt_validation_pipeline/forward_pass_validator.py:268]: Addressed later on in PR #4728 of stack

File: tests/utils/forward_pass_logit_checker.py [hunk 4]
```diff
@@ -690,7 +690,14 @@ def main(config, test_args):  # pylint: disable=W0621
...
       # --- HF Forward Pass ---
       with torch.no_grad():
-        hf_logits_torch = hf_model(**inputs).logits
+        hf_outputs = hf_model(**inputs, output_hidden_states=True)
+        hf_logits_torch = hf_outputs.logits
+        if hasattr(hf_outputs, "hidden_states") and hf_outputs.hidden_states is not None:
+          max_logging.log("--- DUMPING HF INTERMEDIATE ACTIVATIONS ---")
+          max_logging.log(f"Number of layers extracted: {len(hf_outputs.hidden_states)}")
+          for i, layer_tensor in enumerate(hf_outputs.hidden_states):
+            max_logging.log(f"HF Layer {i} Shape: {layer_tensor.shape}, Norm: {torch.norm(layer_tensor, p=2).item():.4f}")
```
[the remark this answers is not in the document]
Comment (olufiyin19) [reply tests/utils/forward_pass_logit_checker.py]: Addressed later on in PR #4728 of stack
